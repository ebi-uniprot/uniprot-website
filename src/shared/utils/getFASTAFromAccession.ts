import entryToFASTAWithHeaders from './entryToFASTAWithHeaders';
import fetchData from './fetchData';

import apiUrls from '../config/apiUrls';

import { UniProtkbAPIModel } from '../../uniprotkb/adapters/uniProtkbConverter';
import { UniParcAPIModel } from '../../uniparc/adapters/uniParcConverter';
import { UniRefAPIModel } from '../../uniref/adapters/uniRefConverter';
import { Namespace } from '../types/namespaces';

const getFASTAFromAccession = async (
  accession?: string
): Promise<string | void> => {
  if (!accession) {
    return;
  }
  let url;
  if (accession.startsWith('UniRef')) {
    url = apiUrls.entry(accession, Namespace.uniref);
  } else if (accession.startsWith('UPI')) {
    url = apiUrls.entry(accession, Namespace.uniparc);
  } else {
    url = apiUrls.entry(accession, Namespace.uniprotkb);
  }
  if (!url) {
    return;
  }
  const response = await fetchData<
    undefined | UniProtkbAPIModel | UniParcAPIModel | UniRefAPIModel
  >(url);
  if (!response.data) {
    return;
  }
  // TODO: handle inactive entries (see P00001) valid response, status code: 200
  // TODO: handle here? or in entryToFASTAWithHeader?
  // eslint-disable-next-line consistent-return
  return entryToFASTAWithHeaders(response.data);
};

export default getFASTAFromAccession;
