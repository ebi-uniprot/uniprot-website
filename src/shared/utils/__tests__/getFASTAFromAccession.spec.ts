import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import getFASTAFromAccession from '../getFASTAFromAccession';

import mockUniProtKB from '../../../uniprotkb/components/__mocks__/swissprotEntry.json';
import mockUniParc from '../../../uniparc/__mocks__/entryModelData.json';
// import mockUniRef from ...

let mock;

beforeAll(() => {
  mock = new MockAdapter(axios);
});

afterEach(() => {
  mock.reset();
});

afterAll(() => {
  mock.restore();
});

describe('getFASTAFromAccession', () => {
  it('should handle no accession', async () => {
    expect(await getFASTAFromAccession()).toBeUndefined();
  });

  it('should not throw if no data', async () => {
    mock.onGet(/api\/uniprotkb\/accession\/P-non-existing$/).reply(204);
    expect(await getFASTAFromAccession('P-non-existing')).toBeUndefined();
  });

  it('should handle UniProtKB entry', async () => {
    mock.onGet(/api\/uniprotkb\/accession\/P05067$/).reply(200, mockUniProtKB);
    expect(await getFASTAFromAccession('P05067')).toMatchSnapshot();
  });

  it('should handle UniParc entry', async () => {
    mock.onGet(/api\/uniparc\/UPI0000000001$/).reply(200, mockUniParc);
    expect(await getFASTAFromAccession('UPI0000000001')).toMatchSnapshot();
  });

  // TODO: uncomment and add a mock json payload once UniRef API has stabilised
  it.skip('should handle UniRef entry', async () => {
    // mock.onGet(/api\/uniref\/......$/).reply(200, mockUniRef);
    // expect(await getFASTAFromAccession('......')).toMatchSnapshot();
  });
});
