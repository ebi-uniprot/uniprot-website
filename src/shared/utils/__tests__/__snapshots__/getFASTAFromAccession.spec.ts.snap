// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFASTAFromAccession should handle UniParc entry 1`] = `
">UPI0000000001 status=true
MGAAASIQTT VNTLSERISS KLEQEANASA QTKCDIEIGN FYIRQNHGCN LTVKNMCSAD
ADAQLDAVLS AATETYSGLT PEQKAYVPAM FTAALNIQTS VNTVVRDFEN YVKQTCNSSA
VVDNKLKIQN VIIDECYGAP GSPTNLEFIN TGSSKGNCAI KALMQLTTKA TTQIAPKQVA
GTGVQFYMIV IGVIILAALF MYYAKRMLFT STNDKIKLIL ANKENVHWTT YMDTFFRTSP
MVIATTDMQN"
`;

exports[`getFASTAFromAccession should handle UniProtKB entry 1`] = `
">sp|P05067|A4_HUMAN Amyloid-beta precursor protein OS=Homo sapiens OX=9606 GN=APP PE=1 SV=3
MLPGLALLLL AAWTARALEV PTDGNAGLLA EPQIAMFCGR LNMHMNVQNG KWDSDPSGTK
TCIDTKEGIL QYCQEVYPEL QITNVVEANQ PVTIQNWCKR GRKQCKTHPH FVIPYRCLVG
EFVSDALLVP DKCKFLHQER MDVCETHLHW HTVAKETCSE KSTNLHDYGM LLPCGIDKFR
GVEFVCCPLA EESDNVDSAD AEEDDSDVWW GGADTDYADG SEDKVVEVAE EEEVAEVEEE
EADDDEDDED GDEVEEEAEE PYEEATERTT SIATTTTTTT ESVEEVVREV CSEQAETGPC
RAMISRWYFD VTEGKCAPFF YGGCGGNRNN FDTEEYCMAV CGSAMSQSLL KTTQEPLARD
PVKLPTTAAS TPDAVDKYLE TPGDENEHAH FQKAKERLEA KHRERMSQVM REWEEAERQA
KNLPKADKKA VIQHFQEKVE SLEQEAANER QQLVETHMAR VEAMLNDRRR LALENYITAL
QAVPPRPRHV FNMLKKYVRA EQKDRQHTLK HFEHVRMVDP KKAAQIRSQV MTHLRVIYER
MNQSLSLLYN VPAVAEEIQD EVDELLQKEQ NYSDDVLANM ISEPRISYGN DALMPSLTET
KTTVELLPVN GEFSLDDLQP WHSFGADSVP ANTENEVEPV DARPAADRGL TTRPGSGLTN
IKTEEISEVK MDAEFRHDSG YEVHHQKLVF FAEDVGSNKG AIIGLMVGGV VIATVIVITL
VMLKKKQYTS IHHGVVEVDA AVTPEERHLS KMQQNGYENP TYKFFEQMQN"
`;

exports[`getFASTAFromAccession should handle UniRef entry 1`] = `
">UniRef100_A0A0B7GQ86 PTS system glucose-specific EIICBA component [includes: glucose permease IIC component glucose-specific phosphotransferase enzyme IIB component glucose-specific phosphotransferase enzyme IIA compon... n=1 Tax=Streptococcus sanguinis TaxID=1305 RepID=A0A0B7GQ86_STRSA
MMKDSFKNIF SFEFWQKFGK ALMVVVAVMP AAGLMISIGK SIPMINPNLG VLVTTGGVLE
QIGWGVIGNL HILFALAIGG SWAKERAGGA FAAGLSFILI NRITGVMFGV TGDMLADKTA
VVKTMFGASI KVSDYFISVL ESPALNMGVF VGIIAGFVGA TAYNKYYNFR KLPDALSFFN
GKRFVPFVVI LRSAIVAIVL SFVWPVVQSG INSFGIWIAN SQDTAPILAP FIYGTLERLL
LPFGLHHMLT IPMNYTELGG VYEVITGSGA GTTVAGQDPL WLAWVTDLVG TKTADPNTYK
HLLETVHPAR FKVGQMIGSF GILMGVAAAI YHNVDADKKH KYKGMMIATA LATFLTGVTE
PIEYMFMFVA TPLYLVYSLV QGAAFAMADI VALRVHSFGS IEFLTRTPMA INAGLGGDII
NFIWVTILFG VVMYFISNFM IKKFNYATPG RNGNYETAEG SDEASSSDST GGKVAAASQA
VNVINLLGGR ANIVDVDACM TRLRVTVKDA EKVGTEEQWK AEGAMGLVMK GQGVQAIYGP
KADVLKSDIQ DLLDSGEVIP ETLPSQKAES EAAEVSYKGV TEEVETVADG QVIDLADVKD
PVFSQKMMGD GFAVEPENGK IYSPVAGTVT SVFPSKHAIG LVTDNGLEVL VHIGLETVSL
EGKPFEVHVS EGQKVAAGDL LVTADLEAIK EAGRETSTIV VFTNAAAIKS VTVEKLGQAS
AKTVVAKVEL"
`;
